File: alegacy/src.old/providers/anthropic.ts
import { ProviderAdapterError, toProviderCallError } from "../open-sse/transport/errors";
import { aggregateCapabilities, capabilitiesOf, createModelCatalog, modelOf } from "../open-sse/transport/catalog";
import type {
  CredentialKind,
  Adapter,
  ProviderCaps,
  ProviderMeta,
  ProviderModel,
  ProviderModelCatalog,
  ProviderOutput,
  ProviderRequest,
  Surface,
  RouteTarget,
} from "../application/contracts";
import type { ProviderCallError } from "../application/contracts";
import { callAnthropicWire } from "../open-sse/transport/protocols/anthropic";

/**
 * Anthropic adapter: the "anthropic-messages" surface over the Messages
 * wire format. Declares explicit prompt caching (cache_control + beta
 * header) and extended thinking.
 */

const ANTHROPIC_SURFACES: readonly Surface[] = ["anthropic-messages"];

const ANTHROPIC_DEFAULT_MODELS: readonly ProviderModel[] = [
  modelOf("claude-opus-4-1", "Claude Opus 4.1", capabilitiesOf({ surfaces: ANTHROPIC_SURFACES, reasoning: true, images: true, search: true, explicitCache: true, promptCacheKey: true })),
  modelOf("claude-sonnet-4-5", "Claude Sonnet 4.5", capabilitiesOf({ surfaces: ANTHROPIC_SURFACES, reasoning: true, images: true, search: true, explicitCache: true, promptCacheKey: true })),
  modelOf("claude-haiku-4-5", "Claude Haiku 4.5", capabilitiesOf({ surfaces: ANTHROPIC_SURFACES, images: true, search: true, explicitCache: true, promptCacheKey: true })),
  modelOf("claude-3-7-sonnet", "Claude 3.7 Sonnet", capabilitiesOf({ surfaces: ANTHROPIC_SURFACES, reasoning: true, images: true, search: true, explicitCache: true, promptCacheKey: true })),
  modelOf("claude-3-5-haiku-latest", "Claude 3.5 Haiku", capabilitiesOf({ surfaces: ANTHROPIC_SURFACES, images: true, search: true, explicitCache: true, promptCacheKey: true })),
];

const ANTHROPIC_FALLBACK_CAPABILITIES: ProviderCaps = capabilitiesOf({
  surfaces: ANTHROPIC_SURFACES,
  reasoning: true,
  images: true,
  search: true,
  explicitCache: true,
  promptCacheKey: true,
});
export interface AnthropicAdapterConfig {
  readonly id?: string;
  readonly displayName?: string;
  readonly baseUrl?: string;
  readonly credentialKind?: CredentialKind;
  readonly auth?: "x-api-key" | "bearer";
  readonly models?: readonly ProviderModel[];
}

export class AnthropicAdapter implements Adapter {
  readonly metadata: ProviderMeta;
  readonly capabilities: ProviderCaps;
  readonly models: ProviderModelCatalog;
  private readonly baseUrl: string;
  private readonly auth: "x-api-key" | "bearer";

  constructor(config: AnthropicAdapterConfig = {}) {
    this.baseUrl = (config.baseUrl ?? "https://api.anthropic.com/v1").replace(/\/+$/, "");
    const models = config.models ?? ANTHROPIC_DEFAULT_MODELS;
    this.models = createModelCatalog(models);
    this.capabilities = aggregateCapabilities(models, ANTHROPIC_FALLBACK_CAPABILITIES);
    this.metadata = {
      id: config.id ?? "anthropic",
      displayName: config.displayName ?? "Anthropic",
      protocol: "anthropic",
      credentialKind: config.credentialKind ?? "api_key",
    };
    this.auth = config.auth ?? "x-api-key";
  }

  resolveTarget(modelId: string, surface: Surface): RouteTarget {
    if (!this.capabilities.surfaces.includes(surface)) {
      throw new ProviderAdapterError({
        kind: "capability_unsupported",
        message: `Provider "${this.metadata.id}" does not support surface "${surface}"`,
        statusCode: 400,
        routeScope: null,
      });
    }
    const __entry = this.models.get(modelId); return { providerId: this.metadata.id, modelId, upstreamModelId: __entry?.upstreamId ?? modelId, surface };
  }
  async call(input: ProviderRequest): Promise<ProviderOutput> {
    this.assertSupported(input);
    const { request, credential } = input;
    const headers: Record<string, string> = {
      "content-type": "application/json",
      accept: request.stream ? "text/event-stream" : "application/json",
      "anthropic-version": "2023-06-01",
    };
    if (this.capabilities.explicitCache && this.capabilities.promptCacheKey) headers["anthropic-beta"] = "prompt-caching-2024-07-31";
    if (this.auth === "bearer") headers.authorization = `Bearer ${credential}`;
    else headers["x-api-key"] = credential;
    return callAnthropicWire(input, this.baseUrl, headers, this.capabilities);
  }


  mapError(error: unknown): ProviderCallError {
    return toProviderCallError(error);
  }


  private assertSupported(input: ProviderRequest): void {
    if (input.target.providerId !== this.metadata.id) {
      throw new ProviderAdapterError({
        kind: "capability_unsupported",
        message: `Adapter "${this.metadata.id}" cannot serve provider "${input.target.providerId}"`,
        statusCode: 400,
        routeScope: null,
      });
    }
    if (!this.capabilities.surfaces.includes(input.target.surface)) {
      throw new ProviderAdapterError({
        kind: "capability_unsupported",
        message: `Provider "${this.metadata.id}" does not support surface "${input.target.surface}"`,
        statusCode: 400,
        routeScope: null,
      });
    }
    if (input.request.stream && !this.capabilities.streaming) {
      throw new ProviderAdapterError({
        kind: "capability_unsupported",
        message: `Provider "${this.metadata.id}" does not support streaming`,
        statusCode: 400,
        routeScope: null,
      });
    }
  }
}

File: router/internal/providers/builtin/anthropic.go
package builtin

func AnthropicAI() ProviderDefinition {
	return ProviderDefinition{
		ID: "anthropic", DisplayName: "Anthropic", Protocol: ProtocolAnthropic, Adapter: AdapterAnthropic, CredentialKind: CredentialAPIKey,
		BaseURL: "https://api.anthropic.com/v1", ModelsDevID: "anthropic",
		Models: []ProviderModel{
			{ID: "claude-3-5-sonnet-20240620", DisplayName: "Claude Sonnet 3.5", ContextWindow: 200000, MaxOutput: 8192, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: false, ToolCalls: false, Images: true}},
			{ID: "claude-3-5-sonnet-20241022", DisplayName: "Claude Sonnet 3.5 v2", ContextWindow: 200000, MaxOutput: 8192, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: false, ToolCalls: false, Images: true}},
			{ID: "claude-3-7-sonnet-20250219", DisplayName: "Claude Sonnet 3.7", ContextWindow: 200000, MaxOutput: 64000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-3-haiku-20240307", DisplayName: "Claude Haiku 3", ContextWindow: 200000, MaxOutput: 4096, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: false, ToolCalls: false, Images: true}},
			{ID: "claude-3-opus-20240229", DisplayName: "Claude Opus 3", ContextWindow: 200000, MaxOutput: 4096, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: false, ToolCalls: false, Images: true}},
			{ID: "claude-3-sonnet-20240229", DisplayName: "Claude Sonnet 3", ContextWindow: 200000, MaxOutput: 4096, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: false, ToolCalls: false, Images: true}},
			{ID: "claude-fable-5", DisplayName: "Claude Fable 5", ContextWindow: 1000000, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-haiku-4-5", DisplayName: "Claude Haiku 4.5", ContextWindow: 200000, MaxOutput: 64000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-haiku-4-5-20251001", DisplayName: "Claude Haiku 4.5", ContextWindow: 200000, MaxOutput: 64000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-mythos-5", DisplayName: "Claude Mythos 5", ContextWindow: 1000000, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-opus-4-0", DisplayName: "Claude Opus 4", ContextWindow: 200000, MaxOutput: 32000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-opus-4-1", DisplayName: "Claude Opus 4.1", ContextWindow: 200000, MaxOutput: 32000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-opus-4-1-20250805", DisplayName: "Claude Opus 4.1", ContextWindow: 200000, MaxOutput: 32000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-opus-4-20250514", DisplayName: "Claude 4 Opus", ContextWindow: 200000, MaxOutput: 32000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: false, ToolCalls: false, Images: true}},
			{ID: "claude-opus-4-5", DisplayName: "Claude Opus 4.5", ContextWindow: 200000, MaxOutput: 64000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-opus-4-5-20251101", DisplayName: "Claude Opus 4.5", ContextWindow: 200000, MaxOutput: 64000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-opus-4-6", DisplayName: "Claude Opus 4.6", ContextWindow: 1000000, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-opus-4-7", DisplayName: "Claude Opus 4.7", ContextWindow: 1000000, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-opus-4-8", DisplayName: "Claude Opus 4.8", ContextWindow: 1000000, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-opus-5", DisplayName: "Claude Opus 5", ContextWindow: 1000000, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-sonnet-4-0", DisplayName: "Claude Sonnet 4", ContextWindow: 200000, MaxOutput: 64000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-sonnet-4-20250514", DisplayName: "Claude 4 Sonnet", ContextWindow: 200000, MaxOutput: 64000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: false, ToolCalls: false, Images: true}},
			{ID: "claude-sonnet-4-5", DisplayName: "Claude Sonnet 4.5", ContextWindow: 1000000, MaxOutput: 64000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-sonnet-4-5-20250929", DisplayName: "Claude Sonnet 4.5", ContextWindow: 1000000, MaxOutput: 64000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-sonnet-4-6", DisplayName: "Claude Sonnet 4.6", ContextWindow: 1000000, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
			{ID: "claude-sonnet-5", DisplayName: "Claude Sonnet 5", ContextWindow: 1000000, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceAnthropicMessages}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		},
	}
}
